File: 0264-ugly-number-ii/0264-ugly-number-ii.cpp
class Solution {
public:
    int nthUglyNumber(int n) {
        unordered_map<int,int>mp;
        vector<int>dp;
        dp.push_back(1);
        if(n==1)
            return 1;
        int i=0,j=0,k=0;
        int count=1;
        while(count<n)
        {
            int mi=min(2*dp[i],min(3*dp[j],5*dp[k]));
           if(mi==2*dp[i])
               i++;
           if(mi==3*dp[j])
               j++;
             if(mi==5*dp[k])
               k++;
            if(mi==1)
            {
                continue;
            }
            else
            {
                dp.push_back(mi);
                mp[mi]=1;
                count++;
            }
        }
        return dp[n-1];
        
    }
};

File: 0264-ugly-number-ii/0264-ugly-number-ii.java
class Solution {
    public int nthUglyNumber(int n) {
        if(n==1){
            return 1;
        }
        PriorityQueue<ArrayList<Long>>uglyNumber = new  PriorityQueue<ArrayList<Long>>(new Comparator<ArrayList<Long>>(){
            public int compare(ArrayList<Long>first,ArrayList<Long> second){
                return first.get(0).compareTo(second.get(0));
            }
        });
        HashMap<Long, Integer> numberExist =  new HashMap<Long, Integer>();
        long [] factors = new long[] {(long)2,(long)3,(long)5};
        uglyNumber.add(new ArrayList(List.of((long)2,(long)1,(long)0)));
        int count=1 ;
        int ans =0;
        while(count<n){
            long currMin= uglyNumber.peek().get(0);
            long originalNumber = uglyNumber.peek().get(1);
            long index = uglyNumber.poll().get(2);
            if(numberExist.getOrDefault(currMin,0)==0){
                numberExist.put(currMin,1);
                uglyNumber.add(new ArrayList<Long>(List.of((long)(currMin*2),(long)currMin,(long)0)));
                count++;
            }
            if(index<2){
                index++;
                // System.out.println(typeof(index));
                long factor = factors[(int)index];
                long newNumber = (long)(originalNumber) * factor;
                uglyNumber.add(new ArrayList<Long>(List.of(newNumber,originalNumber,(long)(index))));
            }
            ans= (int)currMin;
            
        }
        return ans;
        
        
    }
}